### DISCUSSION TITLE
std::map does not automatically deduct types in C++17

### DISCUSSION THREAD

**Discussion by omair-khalid** [cpp-for-yourself/lectures-and-homeworks#28] [General]
When I compile the example code for std::map, the following line gives an error:

`  std::map mutable_map = {std::pair{42, 42.42}, std::pair{23, 23.23}};
`

The error is:
```
class template argument deduction failed:
   std::map mutable_map = {std::pair{42, 42.42}, std::pair{23, 23.23}};
```

I am compiling using c++17, but there still is this error. It gets resolved only when I specify the template types of std::map i.e. `std::map<int, float>  mutable_map = {std::pair{42, 42.42}, std::pair{23, 23.23}};`

Any idea why this is happening to me?

**Comment by jayabrata97** [comment]
Hi @omair-khalid, it is working fine for me. Can you please mention how you are compiling it?

**Reply by omair-khalid** [reply]
Hi Jay, I am compiling it using the following command:
`c++ -std=c++17 -o map map.cpp`

map.cpp is the name of my cpp file.

**Reply by jayabrata97** [reply]
Ok, I am also using the same. Can you please share the complete code? I will also try to compile that.

**Reply by omair-khalid** [reply]
Please find the code below:

```
#include <iostream>
#include <map>
#include <utility>

using std::cout;
using std::endl;

int main() {
  const std::map<int, double> const_map = {{42, 42.42}, {23, 23.23}};
  // Work with a const map
  cout << "Map is empty: " << const_map.empty() << endl;
  cout << "Map size: " << const_map.size() << endl;
  cout << const_map.at(23) << endl;
  // Work with a mutable map
  std::map mutable_map = {std::pair{42, 42.42}, std::pair{23, 23.23}};
  cout << "Map size: " << mutable_map.size() << endl;
  mutable_map[4242] = 23.42; // Creates a new key-value pair
  mutable_map[12] = 15.12;
  cout << "New map size: " << mutable_map.size() << endl;
  cout << mutable_map.at(4242) << endl;
  mutable_map.at(4242) = 42.42; // Ref to an existing value
  cout << mutable_map.at(4242) << endl;
  mutable_map.erase(4242);
  cout << mutable_map.size() << endl;
  return 0;
}
```

Thanks!

**Reply by jayabrata97** [reply]
Hi @omair-khalid, It is working fine for me. I am attaching a screenshot of the code compilation. The code is in transparent background.
![Screenshot from 2022-10-12 23-54-08](https://user-images.githubusercontent.com/78541698/195420685-e5e9c938-1b6b-4844-82d3-16af8c4723e5.png)

**Comment by niosus** [comment]
Hey @omair-khalid, sorry for a late response. For whatever reason, I was not notified of this discussion. Could you also provide some info about the system you're using? Also what does the `c++ --version` print?

**Reply by niosus** [reply]
For completeness, it might be that your compiler is too old and does not fully support all the C++17 features. For example, if I play with your code in the Compiler Explorer, I see that the earliest version of the GCC compiler that can compile this code is 8.1:
- The code compiles with GCC 8.1: https://godbolt.org/z/6bqYb7Trr
- And does not with GCC 7.5: https://godbolt.org/z/q8q5Ynvnv

The same story with `clang`, version 7.0 fails to compile the code, but 8.0 compiles it without issues.

**Reply by jayabrata97** [reply]
Hi @niosus, I have one question: the output on the right-hand side of https://godbolt.org/z/6bqYb7Trr. Is that assembly code?

**Reply by niosus** [reply]
Yep, it is. But it's not important for us for now. We only care if it compiles or not 😉

**Reply by omair-khalid** [reply]
> Hey @omair-khalid, sorry for a late response. For whatever reason, I was not notified of this discussion. Could you also provide some info about the system you're using? Also what does the `c++ --version` print?

Hi Igor, sorry for the delay from my side as well. 
I am using Ubuntu 18. `c++ --version` gives me `c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0`

**Reply by omair-khalid** [reply]
> For completeness, it might be that your compiler is too old and does not fully support all the C++17 features. For example, if I play with your code in the Compiler Explorer, I see that the earliest version of the GCC compiler that can compile this code is 8.1:
> 
> * The code compiles with GCC 8.1: https://godbolt.org/z/6bqYb7Trr
> * And does not with GCC 7.5: https://godbolt.org/z/q8q5Ynvnv
> 
> The same story with `clang`, version 7.0 fails to compile the code, but 8.0 compiles it without issues.

Right! Ill update the compiler and clang to sort it out! Thanks for the help!
